runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -54,6 +56,55 @@ void waitForWorkflow(UUID workflow, ManagementClient client, Duration timeout,
         throw new TimeoutException();
     }
 
+    /**
+     * Remove a node from the cluster.
+     *
+     * @param endpointToRemove Endpoint of the node to be removed from the cluster.
+     * @param retry the number of times to retry a workflow if it fails
+     * @param timeout total time to wait before the workflow times out
+     * @param pollPeriod the poll interval to check whether a workflow completed or not
+     * @throws WorkflowException throws a workflow exception on failure
+     */
+    public void removeNode(@Nonnull String endpointToRemove, int retry,
+                           @Nonnull Duration timeout, @Nonnull Duration pollPeriod) {
+        for (int x = 0; x < retry; x++) {
+            runtime.invalidateLayout();
+            Layout layout = runtime.getLayoutView().getLayout();
+            List<String> layoutServers = new ArrayList(layout.getLayoutServers());
+            layoutServers.remove(endpointToRemove);
+            if (layoutServers.isEmpty()) {
+                throw new WorkflowException("Can't remove node " + endpointToRemove + " from " + layout);
+            }
+
+            String orchestratorEndpoint = layoutServers.get(layoutServers.size() - 1);
+
+            ManagementClient client = runtime.getRouter(orchestratorEndpoint)
+                    .getClient(ManagementClient.class);
+
+            try {
+                CreateWorkflowResponse resp = client.removeNode(endpointToRemove);
+                log.info("removeNode: Workflow id {} for {}", resp.workflowId, endpointToRemove);
+                waitForWorkflow(resp.getWorkflowId(), client, timeout, pollPeriod);
+                Layout newLayout = null;
+                for (int y = 0; y < runtime.getParameters().getInvalidateRetry(); y++) {
+                    runtime.invalidateLayout();
+                    newLayout = runtime.getLayoutView().getLayout();
+                    if (!newLayout.getAllServers().contains(endpointToRemove)) {
+                        log.info("removeNode: Successfully removed {}", endpointToRemove);
+                        return;
+                    }
+                }
+                log.warn("removeNode: Couldn't remove {} from {}, retrying...", endpointToRemove, newLayout, x);
+
+            } catch (NetworkException | TimeoutException  e) {
+                log.warn("removeNode: Error while requesting to remove {}", endpointToRemove, e);
+                continue;
+            }
+        }
+
+        log.warn("removeNode: Failed to remove node {}", endpointToRemove);
```

Comment:
This seems to be incorrect. It is still possible for removeNode to have succeeded at this point. You need to seal the epoch off to ensure removeNode fails at this point.

Reply:
This comment is not clear, the method is called from the application and it is not an atomic operation, why does it need to seal ? This method/exception indicate that the call failed, but it doesn't imply anything about the side effect. I can change the message to "removeNode: call to remove node {} failed"

Reply:
No, the message is not the problem, the problem is that after exiting this method removeNode may have succeeded. You must seal it, otherwise the API is just as ambiguous as before.

Reply:
That's also true for Add node, what if it just times out do you want to seal it too ?

Also, you seal it and freeze the system, then what?

Reply:
The problem is that the API definition itself is ambiguous. Yes, addNode has the same problem too.

If you're going to declare the workflow failed (as per the javadocs), it needs to actually fail. This method doesn't do that. To declare the workflow failed, you need to ensure that the epoch transitions so that the workflow cannot proceed. Sealing (and moving the layout to the previous one) is one way to achieve that.

You can also change the API so that it declares that it can't determine the status of the workflow. I would change the name of the exception to WorkflowResultUnknownException instead in that case.

Reply:
The initial design was based on executing workflows "optimistically", so I'll go with your second suggestion.

Reply:
This comment isn't coming out of left field. Last Friday we discussed (and I thought agreed) that an API which returned ambiguous results was error prone.  And the API can't be optimistic, as that was part of the whole problem in the first place.

Reply:
Note that best effort and optimistic are two different things.

Reply:
One fundamental problem is you can't guarantee the completion of an operation within a specific time period. The first solution that you suggest also doesn't work. You can't guarantee the "rollback" (i.e. seal and restore the previous layout) within a specific time period which results in the same issue. Also, you can have interference by other entities that try to change the layout, so you can't guarantee the layout "rollback" to the same version before the remove node operation started.

In the future we would like to persist the workflow id/status which makes the API much less ambiguous. Not sure how you define optimistic and best effort in this context, maybe you can make your point more explicit.

Reply:
You don't need to guarantee the completion of anything. You just need to make sure the client has an accurate result at the end of the API calls, and that the guarantees are clear. The problem with the past version of this api is that the user could wrongly assume that the operation completed successfully when it did not. The problem with the current version is the opposite, that the user can assume failure when it did not.

Reply:
Added a WorkflowResultUnknownException as you suggested and changed the java doc.